From 42572661e0f0e13687b16b3f6aca16928d3c6622 Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Mon, 8 Oct 2018 17:33:58 +0800
Subject: [PATCH 4871/5242] MLK-19850-4 usb: chipidea: udc: using USB PHY
 charger framework for detection

commit  0fd3d3881e5046c782b71d6f45ae4198f509fc0d from
https://source.codeaurora.org/external/imx/linux-imx.git

Some platforms (eg, imx6/imx7ulp/imx8qm) which implements charger
detection at USB PHY driver can use framework directly. Other
platforms (eg, imx7d/imx845) which do not implement charger detection
at their USB PHY driver, just assign the charger detection results
for usb_phy structure.

Reviewed-by: Jun Li <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/udc.c |   12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/usb/chipidea/udc.c b/drivers/usb/chipidea/udc.c
index 2eb347e..b4e797d 100644
--- a/drivers/usb/chipidea/udc.c
+++ b/drivers/usb/chipidea/udc.c
@@ -1584,10 +1584,10 @@ static int ci_udc_vbus_session(struct usb_gadget *_gadget, int is_active)
 		usb_phy_set_charger_state(ci->usb_phy, is_active ?
 			USB_CHARGER_PRESENT : USB_CHARGER_ABSENT);
 
-	/* Charger Detection */
-	ci_usb_charger_connect(ci, is_active);
-
 	if (ci->usb_phy) {
+		/* Charger Detection */
+		ci_usb_charger_connect(ci, is_active);
+
 		if (is_active)
 			usb_phy_set_event(ci->usb_phy, USB_EVENT_VBUS);
 		else
@@ -2017,7 +2017,10 @@ int ci_usb_charger_connect(struct ci_hdrc *ci, int is_active)
 	if (is_active)
 		pm_runtime_get_sync(ci->dev);
 
-	if (ci->platdata->notify_event) {
+	if (ci->usb_phy->charger_detect) {
+		usb_phy_set_charger_state(ci->usb_phy, is_active ?
+			USB_CHARGER_PRESENT : USB_CHARGER_ABSENT);
+	} else if (ci->platdata->notify_event) {
 		if (is_active)
 			hw_write(ci, OP_USBCMD, USBCMD_RS, 0);
 
@@ -2032,6 +2035,7 @@ int ci_usb_charger_connect(struct ci_hdrc *ci, int is_active)
 			/* Pull down dp */
 			hw_write(ci, OP_USBCMD, USBCMD_RS, 0);
 		}
+		schedule_work(&ci->usb_phy->chg_work);
 	}
 
 	if (!is_active)
-- 
1.7.9.5

